[TSDB] Add support for downsampling `aggregate_metric_double` fields

csoulios opened the pull request:
This PR adds support for downsampling metric fields of type `aggregate_metric_double`, enabling the rollup-of-rollups functionality.

csoulios committed:
Use XContentBuilder for index mapping

csoulios committed:
Merge branch 'main' into tsdb-rollup-of-rollups

csoulios committed:
Merge branch 'main' into tsdb-rollup-of-rollups

csoulios committed:
Added aggregate_metric_double field

to the test index mapping

csoulios committed:
initial support for rollup of rollups

csoulios committed:
Merge branch 'main' into tsdb-rollup-of-rollups

csoulios committed:
Merge branch 'main' into tsdb-rollup-of-rollups

csoulios committed:
Add fetcher support for aggregate_double_metric

Change how we serialize metrics to doc

csoulios committed:
Collect aggregate_metric_double fields

for computing metrics

csoulios committed:
Cleanup

csoulios committed:
Fix gradle dependencies

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/AbstractRollupFieldProducer.java:
```diff
@@ -34,14 +40,14 @@ public String name() {
 
     /**
-     * @return the value of the field.
+     * Resets the collected value to the specific subclass reset value.
```
It might be more descriptive to say "an empty value" - or something like that. "empty" being a fairly vague concept. Or maybe `zero-like value`. I dunno.

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/FieldValueFetcher.java:
```diff
@@ -69,27 +71,39 @@ Object format(Object value) {
+        assert fieldType != null : "Unknown field type for field: [" + field + "]";
+        Map<String, FieldValueFetcher> fetchers = new HashMap<>();
+
+        if (fieldType instanceof AggregateDoubleMetricFieldMapper.AggregateDoubleMetricFieldType aggMetricFieldType) {
```
Is this one of the threads you'd pull on to fix the other `instanceof` checks? If the field type built the fetcher - or if we had a `Map<Class<FieldType>, FetcherFactory>` or something?

nik9000 commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DownsampleActionSingleNodeTests.java:
```diff
@@ -180,47 +181,80 @@ public void setup() {
-                .get()
-        );
+
+        XContentBuilder mapping = jsonBuilder().startObject()
```
oooof this is a bit unreadable. I know you are just reformatting, but can you add blocks or something so I can follow this? Sorry for the trouble.

csoulios replied:
No trouble at all. Fixing this

csoulios committed:
Added testRollupOfRollups test

csoulios committed:
Format test

csoulios committed:
javadoc

csoulios committed:
Merge branch 'main' into tsdb-rollup-of-rollups

csoulios committed:
Merge branch 'main' into tsdb-rollup-of-rollups

csoulios committed:
Create fetcher only if mapped field

exists in the index.

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/FieldValueFetcher.java:
```diff
@@ -69,27 +71,43 @@ Object format(Object value) {
+                }
+            }
+        } else {
+            if (context.fieldExistsInIndex(field)) {
```
Include only the fields that exist in the index, instead of all mapped field. 

This should improve downsampling performance for cases such as ECS schema.

csoulios committed:
Removed fuction that was not used

csoulios committed:
Support aggregate_metric_double label fields

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/FieldValueFetcher.java:
```diff
@@ -55,41 +56,39 @@ FormattedDocValues getLeaf(LeafReaderContext context) {
         return fieldData.load(context).getFormattedValues(format);
     }
 
-    Object format(Object value) {
```
Removing this method because it was not use anywhere. Probably a left-over from old code

csoulios commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DownsampleActionSingleNodeTests.java:
```diff
@@ -137,15 +139,14 @@ protected Collection<Class<? extends Plugin>> getPlugins() {
         return List.of(
             LocalStateCompositeXPackPlugin.class,
             Rollup.class,
-            AnalyticsPlugin.class,
```
This was never used. Probably a left-over. Removing it.

csoulios committed:
cleanup

csoulios committed:
Merge branch 'main' into tsdb-rollup-of-rollups

csoulios committed:
Validate downsampling interval

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/TimeseriesFieldTypeHelper.java:
```diff
@@ -43,28 +40,21 @@ public boolean isTimeSeriesLabel(final String field, final Map<String, ?> unused
         public TimeseriesFieldTypeHelper build(final String timestampField) throws IOException {
-            final MapperService mapperService = indicesService.createIndexMapperServiceForValidation(indexMetadata);
-            final CompressedXContent sourceIndexCompressedXContent = new CompressedXContent(indexMapping);
-            mapperService.merge(MapperService.SINGLE_MAPPING_NAME, sourceIndexCompressedXContent, MapperService.MergeReason.INDEX_TEMPLATE);
```
Was this important?

nik9000 replied:
Ah. It moved.

csoulios replied:
Yes, I moved it one level up so that I can reuse the `MapperService` instance

nik9000 reviewed (commented):
java code looks fine to me. I'd love to plumb the `instanceof` checks through another way, but that's fine for a follow up change. I wish you luck fighting your gradle demons.

csoulios committed:
Add tests for rollups of rollups

csoulios committed:
Use aggregate_metric field mapper as plugin

rollup module extends mapper-aggregate-metric instead of
importing it via `api`. This solves some Classloading
issues

csoulios committed:
Merge branch 'main' into tsdb-rollup-of-rollups

csoulios committed:
Removed unused dependency

from rollup build.gradle file

csoulios commented:
@elasticmachine run elasticsearch-ci/docs

csoulios committed:
Move FieldValueFetcher into server module

Moved FieldValueFetcher in the mappers package so
that it can be retrieved through the MappedFieldType
and remove the instanceof clause

csoulios commented:
@nik9000 I have sorted out the classloading issue. I had to modify the rollup x-pack plugin so that it extends the `x-pack-aggregate-metric` plugin. This solved the problem. For more details see [b39ee34](https://github.com/elastic/elasticsearch/pull/90029/commits/b39ee34020796e0b0bc234c0da76f6a3d1314495)

Finally, I made an effort to move the `FieldValueFetchers` inside the server module so that it is returned by `MappedFieldType`. As discussed, this removes the `instanceof` condtional, but spills the downsmpling logic in the mappers code. To be honest I am not sure what's best (or what's worse). Also, it does not remove the `instanceof` conditionals when creating `MetricFieldProducer` and `LabelFieldProducer` instances. 
The change can be seen in [69664b0](https://github.com/elastic/elasticsearch/pull/90029/commits/69664b0ad94d8cf488131beeb0a4fc2e922bddd9). If you don't like it, I am happy to revert it.

With all those changes, I think the PR is ready for one more review. 
Thanks for all the iterations

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/FieldValueFetcher.java:
```diff
@@ -1,96 +0,0 @@
-/**
- * Utility class used for fetching field values by reading field data
- */
-class FieldValueFetcher {
```
This class became an `interface` and moved into server module in the `org.elasticsearch.index.mapper` package. 

The static methods were moved in the `RollupShardIndexer` class

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java:
```diff
@@ -173,6 +171,36 @@ public RollupIndexerAction.ShardRollupResponse execute() throws IOException {
+        final Map<String, FormattedDocValues> docValuesFetchers = new LinkedHashMap<>(fieldValueFetchers.size());
+        for (FieldValueFetcher fetcher : fieldValueFetchers) {
+            for (Map.Entry<String, FormattedDocValues> e : fetcher.getLeaves(ctx).entrySet()) {
+                if (searchContext.fieldExistsInIndex(e.getKey())) {
```
If field does not exist in the index, it will be ignored even if it exists in the mapping. This should improve performance for use cases such as ECS

nik9000 commented:
> @nik9000 I have sorted out the classloading issue. I had to modify the rollup x-pack plugin so that it extends the x-pack-aggregate-metric plugin. This solved the problem. For more details see [b39ee34](https://github.com/elastic/elasticsearch/pull/90029/commits/b39ee34020796e0b0bc234c0da76f6a3d1314495)

got it. I don't really know if there is a "right" way to do this, but I'm just happy this works.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:
```diff
@@ -110,6 +110,14 @@ public IndexFieldData.Builder fielddataBuilder(FieldDataContext fieldDataContext
+    /**
+     * Create a helper class to fetch value field data.
+     */
+    public FieldValueFetcher fieldValueFetcher(SearchExecutionContext context) {
```
I would prefer to keep this in the rollup module, but this works for now. I can put together a followup PR with some ideas once this is merged. Or I can try. And if I fail then, well, I'll know.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/FieldValueFetcher.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Interface for retrieving field values by reading field data
+ */
+public interface FieldValueFetcher {
```
If it's in `server` it's nice to have a very good name. Its a `FormattedDocValuesFetcher` or something. I dunno. Probably not worth changing it now. If I can't figure out a good way to move it back to the module then we should rename it.

nik9000 reviewed (approved):
LGTM.

I do think it's be nice to keep the fetchers in the rollup module and I think I know how. But let's get this in and I can make a PR and we can see if it's a reasonable way.

csoulios committed:
Revert "Move FieldValueFetcher into server module"

This reverts commit 69664b0ad94d8cf488131beeb0a4fc2e922bddd9.

csoulios commented:
I reverted the last change. I moved the `FieldValueFetcher` class back to the rollup module.I am going to merge this PR as is. Let's work on a following PR on how to remove the `instanceof` conditionals

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc

csoulios committed:
Merge branch 'main' into tsdb-rollup-of-rollups

csoulios commented:
@elasticmachine run elasticsearch-ci/part-1
